### ISSUE TITLE
Input field accessibility

### ISSUE THREAD

**Issue by orangemug** [maputnik/editor#319]
Input elements are not currently associated with their labels. Therefore they are not accessible to a screen reader and break a general UX pattern of not being able to click on a label to focus the input.

There are two ways to achieve this in the DOM

```html
<label for="some-unique-id"></label>
<input id="some-unique-id" />
```

or

```html
<label>
  <input />
</label>
```

**Comment by orangemug** [comment]
So an update with some research, so looking at https://webaim.org/techniques/forms/controls we need to markup our field differently depending on what's in them, examples from https://webaim.org/techniques/forms/controls follow

`<FieldMultiInput/>` should look something like

```html
<fieldset>
  <legend>Translate anchor</legend>
  <label>
    <input type="radio" value="viewport">
    Viewport
  </label>
  <label>
    <input type="radio" value="map">
    Map
  </label>
</fieldset>
```

Whereas `<FieldString/>` should look something like

```html
<label for="id">ID</label>
<input id="id" />
```

The issue is at the moment we have a single container element called `<Block/>` which contains the labels/actions and the children of that component is the <Field* />` components. So for example

```jsx
<Block label={"ID"} fieldSpec={latest.layer.id}
  data-wd-key={this.props.wdKey}
  error={this.props.error}
>
  <FieldString
    value={this.props.value}
    onChange={this.props.onChange}
  />
</Block>
```

I'm going to get rid of the generic `<Block/>` element instead adding in the `fieldSpec`/`error` into the `<Field* />` so for example the above would be rewritten as

```jsx
<FieldString
  label={"ID"}
  fieldSpec={latest.layer.id}
  value={this.props.value}
  onChange={this.props.onChange}
/>
```

Which kind of makes sense because in some `<Field*/>` components we already pass in fieldSpec anyway, for example `<FieldDynamicArray/>`

**Comment by orangemug** [comment]
@pathmapper I've got a little further with this (urgh it's a fair bit of work). Basically the difficulties are that most of the input fields are inaccessible. I think I've fixed most of the issues now and will post a PR hopefully Monday.

There is one aspect that was a little more difficult, that was 'functions', so the issue is that the UI isn't really that self explanatory for a user with perfect sight because the UI is lacking a lot of labels. So I though I'd have a go at redesigning it. Here's what I've come up with

**Before**
<img width="412" alt="Screenshot 2020-06-06 at 18 53 10" src="https://user-images.githubusercontent.com/235915/83951189-ac87e700-a827-11ea-94eb-620b32fd1890.png">

**After**
<img width="361" alt="Screenshot 2020-06-06 at 18 52 50" src="https://user-images.githubusercontent.com/235915/83951199-b4478b80-a827-11ea-96b6-3b04c4e5b14f.png">
Note: It's missing the info/sdk button next to the label at the moment.

The new design is more accessible to a screenreader, there are probably a few things that are still incorrect (I'm still learning). The basic structure is now 

```html
<fieldset>
  <legend>Extrusion height</legend>
  <!-- function, property, default... -->
  <table>
    <!-- stops -->
  </table>
</fieldset>
``` 

I really like this new design with the exception of the 'qualifier' label, I'm not sure from the spec what this should be called.

What do you think? Does this UI make more sense?

**Comment by pathmapper** [comment]
Makes much more sense than before @orangemug :+1: 

> with the exception of the 'qualifier' label, I'm not sure from the spec what this should be called.

https://docs.mapbox.com/mapbox-gl-js/style-spec/other/#function-stops

On your screenshot `12` is the output value.

So what about the following?
- renaming `value` -> `output value`
- renaming `qualifier` -> `input` or `input value`

**Comment by orangemug** [comment]
> So what about the following?
> 
> - renaming Value -> Output Value
> - renaming Qualifier -> Input or Input value

That's much better, thanks @pathmapper

**Comment by orangemug** [comment]
A little update, it's taking a little longer than I planned. I'm refactoring a few things as I go, the working branch is here https://github.com/maputnik/editor/compare/master...orangemug:fix/input-label-a11y

Most of the inputs are a lot more accessible now. Where there are lists in the UI for example (dynamic) arrays of values these are now marked up as lists, so `<ul/>` & `<li/>`.

Labels are now associated with their inputs which makes using keyboard tab-navigation a lot nicer, before you'd get a keyboard focus for the `<label/>` and `<input/>`, now we just get a tab-navigation for the `<input/>`, however the input is still described by the label for screenreaders.  

<img width="955" alt="screenshot of editor with inputs associated with labels" src="https://user-images.githubusercontent.com/235915/84187546-17385d00-aa8a-11ea-9621-60a4d4e61c88.png">


Where we have groups of elements like data/zoom functions or radio buttons they are marked up as fieldsets. We also have the updated UI for both data and zoom functions.

<img width="360" alt="zoom function screenshot" src="https://user-images.githubusercontent.com/235915/84187621-3afba300-aa8a-11ea-8ec9-9d680a61bf79.png">
<img width="361" alt="data function screenshot" src="https://user-images.githubusercontent.com/235915/84187665-4c44af80-aa8a-11ea-87f7-adb4df11ffb9.png">

There are now [storybook](https://storybook.js.org/) stories for most of the components in the UI.

During testing it was becoming a pain point continually having to navigate through the UI on page refresh between changes, although the standalone storybook helped somewhat, when I needed to dev in the actual editor it was always a pain. I've added a search param based router to the editor to help with this. It's of the format

```
http://localhost:8080/editor/?layer=[hash_of_style]~[layer_index]&view=[map_view]&modal=[modals_that_are_open]
```

For example if you select layer index 3, set map view to 'inspect' and open the 'export' modal. The URL would look like the following, where `2995677221` is a hash of the style json.

```
http://localhost:8080/editor/?layer=2995677221~3&view=inspect&modal=export
```

Which means if you refresh the page the UI will return to the state it was previously, which is really speeding up development. It could also be useful for creating some docs in the future as you can link to a UI state. Although that not the priority here, so we might end up disabling it pre-merge to address it (test it) for the next release.

Still left todo

 - Add back in fn, expression, zoom buttons to labels (currently missing)
 - Add back info button and SDK docs to fields
 - Give it all a good test

**Comment by orangemug** [comment]
Closed by #706
